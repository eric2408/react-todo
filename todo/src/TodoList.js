import React, {useState} from "react";
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';


function TodoList() {
    const initial = [];
    const [todos, setTodos] = useState(initial);

    const addTodo = (newTodo) => {
        setTodos(todos => [...todos, newTodo])
    }

    const remove = (id) => {
        setTodos(todos => todos.filter(todo=> todo.id!= id));
    }

    const update = (id, value) => {
        setTodos(todos => todos.map(todo => 
            todo.id === id ? { ...todo, task: value } : todo));
    }

    return (
        <div>
            <h1>My Todo List</h1>
            <NewTodoForm addTodo={addTodo}/>
            <ul>
                {todos.map(todo => 
                <Todo 
                key={todo.id}
                id={todo.id}
                task={todo.task}
                remove={remove}
                update={update}
                />
                )}
            </ul>
        </div>
    );
}

export default TodoList;
