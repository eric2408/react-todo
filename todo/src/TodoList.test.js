import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(todoList, task="testing is part of todo"){
    const form = todoList.getByLabelText('Todo:');
    fireEvent.change(form, { target: { value: task } });
    const btn = todoList.getByText('Add Todo');
    fireEvent.click(btn)
}

it("renders without crashing", function() {
    render(<TodoList />);
  });

it("matches snapshot", function() {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("adding todo to the list works", function(){
    const list = render(<TodoList />);
    addTodo(list);

    expect(list.getByLabelText('Todo:')).toHaveValue("");
    expect(list.getByText("testing is part of todo")).toBeInTheDocument();
    expect(list.getByText("Update")).toBeInTheDocument();
    expect(list.getByText("x")).toBeInTheDocument();
});


it("updating todo list", function(){
    const list = render(<TodoList />);
    addTodo(list);

    fireEvent.click(list.getByText('Update'));
    const updatingInfo = list.getByDisplayValue("testing is part of todo");
    fireEvent.change(updatingInfo, { target: { value: "walk a dog" }});
    fireEvent.click(list.getByText('Update Todo'));

    expect(list.getByText("walk a dog")).toBeInTheDocument();
    expect(list.queryByText("testing is part of todo")).not.toBeInTheDocument();
});

it("deleting a todo from the todo list", function(){
    const list = render(<TodoList />);
    addTodo(list);

    fireEvent.click(list.getByText('x'));

    expect(list.queryByText("testing is part of todo")).not.toBeInTheDocument();
})