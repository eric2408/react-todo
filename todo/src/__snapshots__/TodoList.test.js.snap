// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      My Todo List
    </h1>
    <form>
      <label
        for="task"
      >
        Todo: 
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
